kv:bulk put/delete error

When running:
`wrangler kv:bulk delete --namespace-id=REDACTED keys.json`

I get this error:
```
✘ [ERROR] Received a bad response from the API

  could not parse array of keys from request body: 'invalid character 'a' in literal true (expecting
  'r')' [code: 10012]
  
  If you think this is a bug, please open an issue at:
  https://github.com/cloudflare/wrangler2/issues/new
```

Contents of `keys.json`:
```
[
  "target:foo"
]
```

Similar for `kv:bulk put`.

CLI version: `alpha`

--- a/.changeset/silver-doors-cough.md
+++ b/.changeset/silver-doors-cough.md
@@ -1,0 +1,10 @@
+---
+"wrangler": patch
+---
+
+fix: kv:bulk should JSON encode its contents
+
+The body passed to `kv:bulk delete` and `kv:bulk put` must be JSON encoded.
+This change fixes that and adds some tests to prove it.
+
+Fixes #547

--- a/packages/wrangler/src/__tests__/kv.test.ts
+++ b/packages/wrangler/src/__tests__/kv.test.ts
@@ -1,4 +1,5 @@
 import { writeFileSync } from "node:fs";
+import { Headers } from "undici";
 import { mockAccountId, mockApiToken } from "./helpers/mock-account-id";
 import {
   setMockResponse,
@@ -10,7 +11,7 @@ import { mockConsoleMethods } from "./helpers/mock-console";
 import { mockKeyListRequest } from "./helpers/mock-kv";
 import { runInTempDir } from "./helpers/run-in-tmp";
 import { runWrangler } from "./helpers/run-wrangler";
-import type { KVNamespaceInfo, NamespaceKeyInfo } from "../kv";
+import type { KeyValue, KVNamespaceInfo, NamespaceKeyInfo } from "../kv";
 
 describe("wrangler", () => {
   mockAccountId();
@@ -319,10 +320,7 @@ describe("wrangler", () => {
     describe("put", () => {
       function mockKeyPutRequest(
         expectedNamespaceId: string,
-        expectedKey: string,
-        expectedValue: string,
-        expiration?: number,
-        expirationTtl?: number
+        expectedKV: KeyValue
       ) {
         const requests = { count: 0 };
         setMockResponse(
@@ -332,15 +330,19 @@ describe("wrangler", () => {
             requests.count++;
             expect(accountId).toEqual("some-account-id");
             expect(namespaceId).toEqual(expectedNamespaceId);
-            expect(key).toEqual(expectedKey);
-            expect(body).toEqual(expectedValue);
-            if (expiration !== undefined) {
-              expect(query.get("expiration")).toEqual(`${expiration}`);
+            expect(key).toEqual(expectedKV.key);
+            expect(body).toEqual(expectedKV.value);
+            if (expectedKV.expiration !== undefined) {
+              expect(query.get("expiration")).toEqual(
+                `${expectedKV.expiration}`
+              );
             } else {
               expect(query.has("expiration")).toBe(false);
             }
-            if (expirationTtl) {
-              expect(query.get("expiration_ttl")).toEqual(`${expirationTtl}`);
+            if (expectedKV.expiration_ttl) {
+              expect(query.get("expiration_ttl")).toEqual(
+                `${expectedKV.expiration_ttl}`
+              );
             } else {
               expect(query.has("expiration_ttl")).toBe(false);
             }
@@ -351,11 +353,10 @@ describe("wrangler", () => {
       }
 
       it("should put a key in a given namespace specified by namespace-id", async () => {
-        const requests = mockKeyPutRequest(
-          "some-namespace-id",
-          "my-key",
-          "my-value"
-        );
+        const requests = mockKeyPutRequest("some-namespace-id", {
+          key: "my-key",
+          value: "my-value",
+        });
 
         await runWrangler(
           "kv:key put my-key my-value --namespace-id some-namespace-id"
@@ -370,7 +371,10 @@ describe("wrangler", () => {
 
       it("should put a key in a given namespace specified by binding", async () => {
         writeWranglerConfig();
-        const requests = mockKeyPutRequest("bound-id", "my-key", "my-value");
+        const requests = mockKeyPutRequest("bound-id", {
+          key: "my-key",
+          value: "my-value",
+        });
         await runWrangler(
           "kv:key put my-key my-value --binding someBinding --preview false"
         );
@@ -384,11 +388,10 @@ describe("wrangler", () => {
 
       it("should put a key in a given preview namespace specified by binding", async () => {
         writeWranglerConfig();
-        const requests = mockKeyPutRequest(
-          "preview-bound-id",
-          "my-key",
-          "my-value"
-        );
+        const requests = mockKeyPutRequest("preview-bound-id", {
+          key: "my-key",
+          value: "my-value",
+        });
 
         await runWrangler(
           "kv:key put my-key my-value --binding someBinding --preview"
@@ -402,13 +405,12 @@ describe("wrangler", () => {
       });
 
       it("should add expiration and ttl properties when putting a key", async () => {
-        const requests = mockKeyPutRequest(
-          "some-namespace-id",
-          "my-key",
-          "my-value",
-          10,
-          20
-        );
+        const requests = mockKeyPutRequest("some-namespace-id", {
+          key: "my-key",
+          value: "my-value",
+          expiration: 10,
+          expiration_ttl: 20,
+        });
         await runWrangler(
           "kv:key put my-key my-value --namespace-id some-namespace-id --expiration 10 --ttl 20"
         );
@@ -421,11 +423,10 @@ describe("wrangler", () => {
 
       it("should put a key to the specified environment in a given namespace", async () => {
         writeWranglerConfig();
-        const requests = mockKeyPutRequest(
-          "env-bound-id",
-          "my-key",
-          "my-value"
-        );
+        const requests = mockKeyPutRequest("env-bound-id", {
+          key: "my-key",
+          value: "my-value",
+        });
         await runWrangler(
           "kv:key put my-key my-value --binding someBinding --env some-environment --preview false"
         );
@@ -438,11 +439,10 @@ describe("wrangler", () => {
 
       it("should put a key with a value loaded from a given path", async () => {
         writeFileSync("foo.txt", "file-contents", "utf-8");
-        const requests = mockKeyPutRequest(
-          "some-namespace-id",
-          "my-key",
-          "file-contents"
-        );
+        const requests = mockKeyPutRequest("some-namespace-id", {
+          key: "my-key",
+          value: "file-contents",
+        });
         await runWrangler(
           "kv:key put my-key --namespace-id some-namespace-id --path foo.txt"
         );
@@ -652,11 +652,10 @@ describe("wrangler", () => {
 
       it("should error if a given binding has both preview and non-preview and --preview is not specified", async () => {
         writeWranglerConfig();
-        const requests = mockKeyPutRequest(
-          "preview-bound-id",
-          "my-key",
-          "my-value"
-        );
+        const requests = mockKeyPutRequest("preview-bound-id", {
+          key: "my-key",
+          value: "my-value",
+        });
         await expect(
           runWrangler("kv:key put my-key my-value --binding someBinding")
         ).rejects.toThrowErrorMatchingInlineSnapshot(
@@ -1117,6 +1116,156 @@ describe("wrangler", () => {
           `kv:key delete --binding someBinding --env some-environment --preview someKey`
         );
         expect(requests.count).toEqual(1);
+      });
+    });
+  });
+
+  describe("kv:bulk", () => {
+    describe("put", () => {
+      function mockPutRequest(
+        expectedNamespaceId: string,
+        expectedKeyValues: KeyValue[]
+      ) {
+        const requests = { count: 0 };
+        setMockResponse(
+          "/accounts/:accountId/storage/kv/namespaces/:namespaceId/bulk",
+          "PUT",
+          ([_url, accountId, namespaceId], { body }) => {
+            requests.count++;
+            expect(accountId).toEqual("some-account-id");
+            expect(namespaceId).toEqual(expectedNamespaceId);
+            expect(JSON.parse(body as string)).toEqual(expectedKeyValues);
+            return null;
+          }
+        );
+        return requests;
+      }
+
+      it("should put the key-values parsed from a file", async () => {
+        const keyValues: KeyValue[] = [
+          { key: "someKey1", value: "someValue1" },
+          { key: "ns:someKey2", value: "123", base64: true },
+          { key: "someKey3", value: "someValue3", expiration: 100 },
+          { key: "someKey4", value: "someValue4", expiration_ttl: 500 },
+        ];
+        writeFileSync("./keys.json", JSON.stringify(keyValues));
+        const requests = mockPutRequest("some-namespace-id", keyValues);
+        await runWrangler(
+          `kv:bulk put --namespace-id some-namespace-id keys.json`
+        );
+        expect(requests.count).toEqual(1);
+      });
+
+      it("should error if the file is not a JSON array", async () => {
+        const keyValues = { key: "someKey1", value: "someValue1" };
+        writeFileSync("./keys.json", JSON.stringify(keyValues));
+        await expect(
+          runWrangler(`kv:bulk put --namespace-id some-namespace-id keys.json`)
+        ).rejects.toThrowErrorMatchingInlineSnapshot(`
+                "Unexpected JSON input from \\"keys.json\\".
+                Expected an array of key-value objects but got type \\"object\\"."
+              `);
+      });
+
+      it("should error if the array contains items that are not key-value objects", async () => {
+        const keyValues = [
+          123,
+          "a string",
+          { key: "someKey" },
+          { value: "someValue" },
+          { key: "someKey1", value: "someValue1", invalid: true },
+        ];
+        writeFileSync("./keys.json", JSON.stringify(keyValues));
+        await expect(
+          runWrangler(`kv:bulk put --namespace-id some-namespace-id keys.json`)
+        ).rejects.toThrowErrorMatchingInlineSnapshot(`
+                "Unexpected JSON input from \\"keys.json\\".
+                Each item in the array should be an object that matches:
+
+                interface KeyValue {
+                  key: string;
+                  value: string;
+                  expiration?: number;
+                  expiration_ttl?: number;
+                  metadata?: object;
+                  base64?: boolean;
+                }
+
+                The item at index 0 is type: \\"number\\" - 123
+                The item at index 1 is type: \\"string\\" - \\"a string\\"
+                The item at index 2 is {\\"key\\":\\"someKey\\"}
+                The item at index 3 is {\\"value\\":\\"someValue\\"}"
+              `);
+
+        expect(std.out).toMatchInlineSnapshot(`""`);
+        expect(std.warn).toMatchInlineSnapshot(`
+          "Unexpected key-value properties in \\"keys.json\\".
+          The item at index 4 contains unexpected properties: [\\"invalid\\"]."
+        `);
+      });
+    });
+
+    describe("delete", () => {
+      function mockDeleteRequest(
+        expectedNamespaceId: string,
+        expectedKeys: string[]
+      ) {
+        const requests = { count: 0 };
+        setMockResponse(
+          "/accounts/:accountId/storage/kv/namespaces/:namespaceId/bulk",
+          "DELETE",
+          ([_url, accountId, namespaceId], { headers, body }) => {
+            requests.count++;
+            expect(accountId).toEqual("some-account-id");
+            expect(namespaceId).toEqual(expectedNamespaceId);
+            expect(new Headers(headers ?? []).get("Content-Type")).toEqual(
+              "application/json"
+            );
+            expect(JSON.parse(body as string)).toEqual(expectedKeys);
+            return null;
+          }
+        );
+        return requests;
+      }
+
+      it("should delete the keys parsed from a file", async () => {
+        const keys = ["someKey1", "ns:someKey2"];
+        writeFileSync("./keys.json", JSON.stringify(keys));
+        const requests = mockDeleteRequest("some-namespace-id", keys);
+        await runWrangler(
+          `kv:bulk delete --namespace-id some-namespace-id keys.json`
+        );
+        expect(requests.count).toEqual(1);
+      });
+
+      it("should error if the file is not a JSON array", async () => {
+        const keys = 12354;
+        writeFileSync("./keys.json", JSON.stringify(keys));
+        await expect(
+          runWrangler(
+            `kv:bulk delete --namespace-id some-namespace-id keys.json`
+          )
+        ).rejects.toThrowErrorMatchingInlineSnapshot(`
+                "Unexpected JSON input from \\"keys.json\\".
+                Expected an array of strings but got:
+                12354"
+              `);
+      });
+
+      it("should error if the file contains non-string items", async () => {
+        const keys = ["good", 12354, { key: "someKey" }, null];
+        writeFileSync("./keys.json", JSON.stringify(keys));
+        await expect(
+          runWrangler(
+            `kv:bulk delete --namespace-id some-namespace-id keys.json`
+          )
+        ).rejects.toThrowErrorMatchingInlineSnapshot(`
+                "Unexpected JSON input from \\"keys.json\\".
+                Expected an array of strings.
+                The item at index 1 is a number: 12354
+                The item at index 2 is a object: [object Object]
+                The item at index 3 is a object: null"
+              `);
       });
     });
   });

--- a/packages/wrangler/src/index.tsx
+++ b/packages/wrangler/src/index.tsx
@@ -26,6 +26,8 @@ import {
   createNamespace,
   isValidNamespaceBinding,
   getKeyValue,
+  isKeyValue,
+  unexpectedKeyValueProps,
 } from "./kv";
 import { getPackageManager } from "./package-manager";
 import { pages } from "./pages";
@@ -1864,7 +1866,9 @@ export async function main(argv: string[]): Promise<void> {
             }
             // -- snip, end --
 
-            await putKeyValue(config.account_id, namespaceId, key, value, {
+            await putKeyValue(config.account_id, namespaceId, {
+              key,
+              value,
               expiration,
               expiration_ttl: ttl,
             });
@@ -2089,6 +2093,60 @@ export async function main(argv: string[]): Promise<void> {
             const namespaceId = getNamespaceId(args, config);
             const content = parseJSON(readFileSync(filename), filename);
 
+            if (!Array.isArray(content)) {
+              throw new Error(
+                `Unexpected JSON input from "${filename}".\n` +
+                  `Expected an array of key-value objects but got type "${typeof content}".`
+              );
+            }
+
+            const errors: string[] = [];
+            const warnings: string[] = [];
+            for (let i = 0; i < content.length; i++) {
+              const keyValue = content[i];
+              if (typeof keyValue !== "object") {
+                errors.push(
+                  `The item at index ${i} is type: "${typeof keyValue}" - ${JSON.stringify(
+                    keyValue
+                  )}`
+                );
+              } else if (!isKeyValue(keyValue)) {
+                errors.push(
+                  `The item at index ${i} is ${JSON.stringify(keyValue)}`
+                );
+              } else {
+                const props = unexpectedKeyValueProps(keyValue);
+                if (props.length > 0) {
+                  warnings.push(
+                    `The item at index ${i} contains unexpected properties: ${JSON.stringify(
+                      props
+                    )}.`
+                  );
+                }
+              }
+            }
+            if (warnings.length > 0) {
+              console.warn(
+                `Unexpected key-value properties in "${filename}".\n` +
+                  warnings.join("\n")
+              );
+            }
+            if (errors.length > 0) {
+              throw new Error(
+                `Unexpected JSON input from "${filename}".\n` +
+                  `Each item in the array should be an object that matches:\n\n` +
+                  `interface KeyValue {\n` +
+                  `  key: string;\n` +
+                  `  value: string;\n` +
+                  `  expiration?: number;\n` +
+                  `  expiration_ttl?: number;\n` +
+                  `  metadata?: object;\n` +
+                  `  base64?: boolean;\n` +
+                  `}\n\n` +
+                  errors.join("\n")
+              );
+            }
+
             // -- snip, extract --
             const loggedIn = await loginOrRefreshIfRequired();
             if (!loggedIn) {
@@ -2141,7 +2199,34 @@ export async function main(argv: string[]): Promise<void> {
           async ({ filename, ...args }) => {
             const config = readConfig(args.config as ConfigPath);
             const namespaceId = getNamespaceId(args, config);
-            const content = parseJSON(readFileSync(filename), filename);
+            const content = parseJSON(
+              readFileSync(filename),
+              filename
+            ) as string[];
+
+            if (!Array.isArray(content)) {
+              throw new Error(
+                `Unexpected JSON input from "${filename}".\n` +
+                  `Expected an array of strings but got:\n${content}`
+              );
+            }
+
+            const errors: string[] = [];
+            for (let i = 0; i < content.length; i++) {
+              const key = content[i];
+              if (typeof key !== "string") {
+                errors.push(
+                  `The item at index ${i} is a ${typeof key}: ${key}`
+                );
+              }
+            }
+            if (errors.length > 0) {
+              throw new Error(
+                `Unexpected JSON input from "${filename}".\n` +
+                  `Expected an array of strings.\n` +
+                  errors.join("\n")
+              );
+            }
 
             // -- snip, extract --
             const loggedIn = await loginOrRefreshIfRequired();

--- a/packages/wrangler/src/kv.tsx
+++ b/packages/wrangler/src/kv.tsx
@@ -90,26 +90,64 @@ export async function listNamespaceKeys(
   );
 }
 
+/**
+ * Information about a key-value pair, including its "metadata" fields.
+ */
+export interface KeyValue {
+  key: string;
+  value: string;
+  expiration?: number;
+  expiration_ttl?: number;
+  metadata?: object;
+  base64?: boolean;
+}
+
+const KeyValueKeys = new Set([
+  "key",
+  "value",
+  "expiration",
+  "expiration_ttl",
+  "metadata",
+  "base64",
+]);
+
+/**
+ * Is the given object a valid `KeyValue` type?
+ */
+export function isKeyValue(keyValue: object): keyValue is KeyValue {
+  const props = Object.keys(keyValue);
+  if (!props.includes("key") || !props.includes("value")) {
+    return false;
+  }
+  return true;
+}
+
+/**
+ * Get all the properties on the `keyValue` that are not expected.
+ */
+export function unexpectedKeyValueProps(keyValue: KeyValue): string[] {
+  const props = Object.keys(keyValue);
+  return props.filter((prop) => !KeyValueKeys.has(prop));
+}
+
 export async function putKeyValue(
   accountId: string,
   namespaceId: string,
-  key: string,
-  value: string,
-  args?: { expiration?: number; expiration_ttl?: number }
+  keyValue: KeyValue
 ) {
   let searchParams: URLSearchParams | undefined;
-  if (args) {
+  if (keyValue.expiration || keyValue.expiration_ttl) {
     searchParams = new URLSearchParams();
-    if (args.expiration) {
-      searchParams.set("expiration", `${args.expiration}`);
+    if (keyValue.expiration) {
+      searchParams.set("expiration", `${keyValue.expiration}`);
     }
-    if (args.expiration_ttl) {
-      searchParams.set("expiration_ttl", `${args.expiration_ttl}`);
+    if (keyValue.expiration_ttl) {
+      searchParams.set("expiration_ttl", `${keyValue.expiration_ttl}`);
     }
   }
   return await fetchResult(
-    `/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/values/${key}`,
-    { method: "PUT", body: value },
+    `/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/values/${keyValue.key}`,
+    { method: "PUT", body: keyValue.value },
     searchParams
   );
 }
@@ -125,13 +163,13 @@ export async function getKeyValue(
 export async function putBulkKeyValue(
   accountId: string,
   namespaceId: string,
-  keyvalueStr: string
+  keyValues: KeyValue[]
 ) {
   return await fetchResult(
     `/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/bulk`,
     {
       method: "PUT",
-      body: keyvalueStr,
+      body: JSON.stringify(keyValues),
       headers: { "Content-Type": "application/json" },
     }
   );
@@ -140,13 +178,13 @@ export async function putBulkKeyValue(
 export async function deleteBulkKeyValue(
   accountId: string,
   namespaceId: string,
-  keyStr: string
+  keys: string[]
 ) {
   return await fetchResult(
     `/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/bulk`,
     {
       method: "DELETE",
-      body: keyStr,
+      body: JSON.stringify(keys),
       headers: { "Content-Type": "application/json" },
     }
   );

--- a/packages/wrangler/src/sites.tsx
+++ b/packages/wrangler/src/sites.tsx
@@ -9,6 +9,7 @@ import {
   putBulkKeyValue,
 } from "./kv";
 import type { Config } from "./config";
+import type { KeyValue } from "./kv";
 import type { XXHashAPI } from "xxhash-wasm";
 
 /** Paths to always ignore. */
@@ -128,11 +129,7 @@ export async function syncAssets(
   const keys = new Set(result.map((x) => x.name));
 
   const manifest: Record<string, string> = {};
-  const upload: {
-    key: string;
-    value: string;
-    base64: boolean;
-  }[] = [];
+  const upload: KeyValue[] = [];
 
   const include = createPatternMatcher(siteAssets.includePatterns, false);
   const exclude = createPatternMatcher(siteAssets.excludePatterns, true);
@@ -167,7 +164,7 @@ export async function syncAssets(
     }
     manifest[path.relative(siteAssets.baseDirectory, file)] = assetKey;
   }
-  await putBulkKeyValue(accountId, namespace, JSON.stringify(upload));
+  await putBulkKeyValue(accountId, namespace, upload);
   return { manifest, namespace };
 }
 

